### PR TITLE
MCO-1125: OCPBUGS-35277: Allow paths to be defined for non-disruptive updates

### PR DESCRIPTION
**Pull request by djoshy**
**- What I did**

- Add support for paths in file based NodeDisruptionPolicies. If multiple policies apply to a file being updated, the closest match will be considered.
- Moved some of the "backdoor" directory based exceptions into the MCO's default policies.
- Added a new default policy for the user CA cert to maintain feature parity until it moves to cert_writer.
- Updated docs to reflect new defaults and path support.

**- How to verify it**
- All previous NodeDisruptionPolicy behavior should still behave the same(node disruption e2e should pass)
- The daemon should now behave correctly for user CA cert changes instead of rebooting.
- To test the multi-path based policies:

1. Define more than one file based policy:
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
  nodeDisruptionPolicy:
    files:
      - path: /etc/example/test
        actions:
          - type: None
      - path: /etc/example/
        actions:
          - reload:
              serviceName: crio.service
            type: Reload
```
2. Create an MC targeting the infra/worker pool that adds a file under /etc/example. Observe the daemon logs, this should cause a CRIO reload.
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: infra
  name:  111-test-file
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,SSBhbSB2ZXJzaW9uIDEKasas
        path: /etc/example/file
        user:
          name: core
```

3. Now create another MC targeting the same pool that adds a file under /etc/example/test. Observe the daemon logs, this should not cause any action.
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: infra
  name:  112-test-file
spec:
  config:
    ignition:
      version: 3.2.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,SSBhbSB2ZXJzaW9uIDEKasas
        path: /etc/example/test/file
        user:
          name: core
```
**Note: Be sure to add a new MC, and not to overwrite the previous one from step (2). If you overwrite the first test MC, it will cause both policies to be in effect, as you are removing the old test file.**

4. That's it! Nice job! :rocket:

### CODE REVIEW

**Comment by sergiordlr** [comment]
Verified using IPI on AWS

Test cases passed with enabled techpreview:

```
"[sig-mco] MCO security Author:rioliu-NonHyperShiftHOST-NonPreRelease-Longduration-High-71991-post action of user-ca-bundle change will skip drain,reboot and restart crio service [Disruptive] [Serial]"
"[sig-mco] MCO security Author:sregidor-NonHyperShiftHOST-High-67660-MCS generates ignition configs with certs [Disruptive] [Serial]"
```

/label qe-approved

**Comment by sergiordlr** [comment]
/hold 

Still verifying the path functionality

**Comment by sergiordlr** [comment]
Verified path support using IPI on AWS

Configure 3 paths in the Machineconfiguration nodedisruption section

- (dir) /etc/test-file-policy-subdir-75109/extradir/ 
    action none 
- (dir) /etc/test-file-policy-subdir-75109
     action restart crio
- (file) /etc/test-file-policy-subdir-75109/test-file.txt
     action reload crio


1 Create a MC for /etc/test-file-policy-subdir-75109/test-file.txt
2 Check that  crio.service was reloaded
3 Create a MC for /etc/test-file-policy-subdir-75109/test-dir.txt
4 Check that crio.service was restarted
5 Create a MC for  /etc/test-file-policy-subdir-75109/extradir/test-file.txt
6 Check that no action is executed

The commands for every step can be found in test case: [OCP-75109 - NodeDisruptionPolicy files allow paths to be defined for non-disruptive updates](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-75109)


/remove-label qe-approved

**Review by yuqi-zhang — APPROVED** [review]
General logic makes sense to me! I think 2 things here:
1. let's add some additional unit testing around `FindClosestFilePolicyPathMatch`. I don't think there should be any edge cases but best to be safe, especially if you have multiple policies
2. let's make sure we capture this change in other documentation as well (maybe enhancement? for sure the openshift docs)

**Comment by djoshy** [comment]
> let's add some additional unit testing around FindClosestFilePolicyPathMatch. I don't think there should be any edge cases but best to be safe, especially if you have multiple policies

Added a few cases I could think of, let me know if that works!

> let's make sure we capture this change in other documentation as well (maybe enhancement? for sure the openshift docs)

I'll make a card in the epic to track this (:

File: pkg/daemon/update_test.go [hunk 1]
```diff
@@ -707,3 +708,123 @@ func TestIsImagePresent(t *testing.T) {
...
+		"ReloadCrio":  {{Type: opv1.ReloadStatusAction, Reload: &opv1.ReloadService{ServiceName: "crio.service"}}}}
+
+	tests := []struct {
+		diffPath             string
+		filePolicies         []opv1.NodeDisruptionPolicyStatusFile
+		expectedPathFound    bool
+		expectedActionsFound []opv1.NodeDisruptionPolicyStatusAction
+	}{
+		{
+			// test that an file with no polcies returns no actions
```
Comment (yuqi-zhang) [inline pkg/daemon/update_test.go:728]: typo: `polcies` -> `policies`

**Review by yuqi-zhang — COMMENTED** [review]
/lgtm

Tests lgtm, thanks!

File: pkg/daemon/update_test.go (hunk 1 above)
Comment (djoshy) [reply pkg/daemon/update_test.go]: whoops, fixed!

**Pull request merged**